File: packages/code-storage-typescript/src/diff-commit.ts
import { buildCommitResult, parseCommitPackError } from './commit-pack';
import { RefUpdateError } from './errors';
import type { CommitPackAckRaw } from './schemas';
import { commitPackAckSchema } from './schemas';
import {
  type ChunkSegment,
  base64Encode,
  chunkify,
  requiresDuplex,
  toAsyncIterable,
  toRequestBody,
} from './stream-utils';
import type {
  CommitResult,
  CommitSignature,
  CreateCommitFromDiffOptions,
  DiffSource,
} from './types';
import { getUserAgent } from './version';

interface DiffCommitMetadataPayload {
  target_branch: string;
  expected_head_sha?: string;
  base_branch?: string;
  commit_message: string;
  ephemeral?: boolean;
  ephemeral_base?: boolean;
  author: {
    name: string;
    email: string;
  };
  committer?: {
    name: string;
    email: string;
  };
}

interface DiffCommitTransportRequest {
  authorization: string;
  signal?: AbortSignal;
  metadata: DiffCommitMetadataPayload;
  diffChunks: AsyncIterable<ChunkSegment>;
}

interface DiffCommitTransport {
  send(request: DiffCommitTransportRequest): Promise<CommitPackAckRaw>;
}

type NormalizedDiffCommitOptions = {
  targetBranch: string;
  commitMessage: string;
  expectedHeadSha?: string;
  baseBranch?: string;
  ephemeral?: boolean;
  ephemeralBase?: boolean;
  author: CommitSignature;
  committer?: CommitSignature;
  signal?: AbortSignal;
  ttl?: number;
  initialDiff: DiffSource;
};

interface CommitFromDiffSendDeps {
  options: CreateCommitFromDiffOptions;
  getAuthToken: () => Promise<string>;
  transport: DiffCommitTransport;
}

class DiffCommitExecutor {
  private readonly options: NormalizedDiffCommitOptions;
  private readonly getAuthToken: () => Promise<string>;
  private readonly transport: DiffCommitTransport;
  private readonly diffFactory: () => AsyncIterable<Uint8Array>;
  private sent = false;

  constructor(deps: CommitFromDiffSendDeps) {
    this.options = normalizeDiffCommitOptions(deps.options);
    this.getAuthToken = deps.getAuthToken;
    this.transport = deps.transport;

    const trimmedMessage = this.options.commitMessage?.trim();
    const trimmedAuthorName = this.options.author?.name?.trim();
    const trimmedAuthorEmail = this.options.author?.email?.trim();

    if (!trimmedMessage) {
      throw new Error('createCommitFromDiff commitMessage is required');
    }
    if (!trimmedAuthorName || !trimmedAuthorEmail) {
      throw new Error(
        'createCommitFromDiff author name and email are required'
      );
    }

    this.options.commitMessage = trimmedMessage;
    this.options.author = {
      name: trimmedAuthorName,
      email: trimmedAuthorEmail,
    };

    if (typeof this.options.expectedHeadSha === 'string') {
      this.options.expectedHeadSha = this.options.expectedHeadSha.trim();
    }
    if (typeof this.options.baseBranch === 'string') {
      const trimmedBase = this.options.baseBranch.trim();
      if (trimmedBase === '') {
        delete this.options.baseBranch;
      } else {
        if (trimmedBase.startsWith('refs/')) {
          throw new Error(
            'createCommitFromDiff baseBranch must not include refs/ prefix'
          );
        }
        this.options.baseBranch = trimmedBase;
      }
    }
    if (this.options.ephemeralBase && !this.options.baseBranch) {
      throw new Error('createCommitFromDiff ephemeralBase requires baseBranch');
    }

    this.diffFactory = () => toAsyncIterable(this.options.initialDiff);
  }

  async send(): Promise<CommitResult> {
    this.ensureNotSent();
    this.sent = true;

    const metadata = this.buildMetadata();
    const diffIterable = chunkify(this.diffFactory());

    const authorization = await this.getAuthToken();
    const ack = await this.transport.send({
      authorization,
      signal: this.options.signal,
      metadata,
      diffChunks: diffIterable,
    });

    return buildCommitResult(ack);
  }

  private buildMetadata(): DiffCommitMetadataPayload {
    const metadata: DiffCommitMetadataPayload = {
      target_branch: this.options.targetBranch,
      commit_message: this.options.commitMessage,
      author: {
        name: this.options.author.name,
        email: this.options.author.email,
      },
    };

    if (this.options.expectedHeadSha) {
      metadata.expected_head_sha = this.options.expectedHeadSha;
    }
    if (this.options.baseBranch) {
      metadata.base_branch = this.options.baseBranch;
    }
    if (this.options.committer) {
      metadata.committer = {
        name: this.options.committer.name,
        email: this.options.committer.email,
      };
    }
    if (this.options.ephemeral) {
      metadata.ephemeral = true;
    }
    if (this.options.ephemeralBase) {
      metadata.ephemeral_base = true;
    }

    return metadata;
  }

  private ensureNotSent(): void {
    if (this.sent) {
      throw new Error('createCommitFromDiff cannot be reused after send()');
    }
  }
}

export class FetchDiffCommitTransport implements DiffCommitTransport {
  private readonly url: string;

  constructor(config: { baseUrl: string; repoId: string }) {
    const trimmedBase = config.baseUrl.replace(/\/+$/, '');
    this.url = `${trimmedBase}/api/repos/${encodeURIComponent(config.repoId)}/diff-commit`;
  }

  async send(request: DiffCommitTransportRequest): Promise<CommitPackAckRaw> {
    const bodyIterable = buildMessageIterable(
      request.metadata,
      request.diffChunks
    );
    const body = toRequestBody(bodyIterable);

    const init: RequestInit = {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${request.authorization}`,
        'Content-Type': 'application/x-ndjson',
        Accept: 'application/json',
        'Code-Storage-Agent': getUserAgent(),
      },
      body: body as any,
      signal: request.signal,
    };

    if (requiresDuplex(body)) {
      (init as RequestInit & { duplex: 'half' }).duplex = 'half';
    }

    const response = await fetch(this.url, init);
    if (!response.ok) {
      const fallbackMessage = `createCommitFromDiff request failed (${response.status} ${response.statusText})`;
      const { statusMessage, statusLabel, refUpdate } =
        await parseCommitPackError(response, fallbackMessage);
      throw new RefUpdateError(statusMessage, {
        status: statusLabel,
        message: statusMessage,
        refUpdate,
      });
    }

    return commitPackAckSchema.parse(await response.json());
  }
}

function buildMessageIterable(
  metadata: DiffCommitMetadataPayload,
  diffChunks: AsyncIterable<ChunkSegment>
): AsyncIterable<Uint8Array> {
  const encoder = new TextEncoder();
  return {
    async *[Symbol.asyncIterator]() {
      yield encoder.encode(`${JSON.stringify({ metadata })}\n`);
      for await (const segment of diffChunks) {
        const payload = {
          diff_chunk: {
            data: base64Encode(segment.chunk),
            eof: segment.eof,
          },
        };
        yield encoder.encode(`${JSON.stringify(payload)}\n`);
      }
    },
  };
}

function normalizeDiffCommitOptions(
  options: CreateCommitFromDiffOptions
): NormalizedDiffCommitOptions {
  if (!options || typeof options !== 'object') {
    throw new Error('createCommitFromDiff options are required');
  }

  if (options.diff === undefined || options.diff === null) {
    throw new Error('createCommitFromDiff diff is required');
  }

  const targetBranch = normalizeBranchName(options.targetBranch);

  let committer: CommitSignature | undefined;
  if (options.committer) {
    const name = options.committer.name?.trim();
    const email = options.committer.email?.trim();
    if (!name || !email) {
      throw new Error(
        'createCommitFromDiff committer name and email are required when provided'
      );
    }
    committer = { name, email };
  }

  return {
    targetBranch,
    commitMessage: options.commitMessage,
    expectedHeadSha: options.expectedHeadSha,
    baseBranch: options.baseBranch,
    ephemeral: options.ephemeral === true,
    ephemeralBase: options.ephemeralBase === true,
    author: options.author,
    committer,
    signal: options.signal,
    ttl: options.ttl,
    initialDiff: options.diff,
  };
}

function normalizeBranchName(value: string | undefined): string {
  const trimmed = value?.trim();
  if (!trimmed) {
    throw new Error('createCommitFromDiff targetBranch is required');
  }
  if (trimmed.startsWith('refs/heads/')) {
    const branch = trimmed.slice('refs/heads/'.length).trim();
    if (!branch) {
      throw new Error(
        'createCommitFromDiff targetBranch must include a branch name'
      );
    }
    return branch;
  }
  if (trimmed.startsWith('refs/')) {
    throw new Error(
      'createCommitFromDiff targetBranch must not include refs/ prefix'
    );
  }
  return trimmed;
}

export async function sendCommitFromDiff(
  deps: CommitFromDiffSendDeps
): Promise<CommitResult> {
  const executor = new DiffCommitExecutor(deps);
  return executor.send();
}

File: packages/code-storage-go/diff_commit.go
package storage

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type diffCommitExecutor struct {
	options CommitFromDiffOptions
	client  *Client
}

func (d *diffCommitExecutor) normalize() (CommitFromDiffOptions, error) {
	options := d.options
	options.TargetBranch = strings.TrimSpace(options.TargetBranch)
	options.CommitMessage = strings.TrimSpace(options.CommitMessage)
	options.ExpectedTargetSHA = preferredString(options.ExpectedTargetSHA, options.ExpectedHeadSHA)
	options.TargetIsEphemeral = preferredBool(options.TargetIsEphemeral, options.Ephemeral)
	options.BaseIsEphemeral = preferredBool(options.BaseIsEphemeral, options.EphemeralBase)
	options.BaseBranch = strings.TrimSpace(options.BaseBranch)

	if options.Diff == nil {
		return options, errors.New("createCommitFromDiff diff is required")
	}

	branch, err := normalizeDiffBranchName(options.TargetBranch)
	if err != nil {
		return options, err
	}
	options.TargetBranch = branch

	if options.CommitMessage == "" {
		return options, errors.New("createCommitFromDiff commitMessage is required")
	}

	if strings.TrimSpace(options.Author.Name) == "" || strings.TrimSpace(options.Author.Email) == "" {
		return options, errors.New("createCommitFromDiff author name and email are required")
	}
	options.Author.Name = strings.TrimSpace(options.Author.Name)
	options.Author.Email = strings.TrimSpace(options.Author.Email)

	if options.BaseBranch != "" && strings.HasPrefix(options.BaseBranch, "refs/") {
		return options, errors.New("createCommitFromDiff baseBranch must not include refs/ prefix")
	}
	if options.BaseIsEphemeral != nil && *options.BaseIsEphemeral && options.BaseBranch == "" {
		return options, errors.New("createCommitFromDiff baseIsEphemeral requires baseBranch")
	}

	if options.Committer != nil {
		if strings.TrimSpace(options.Committer.Name) == "" || strings.TrimSpace(options.Committer.Email) == "" {
			return options, errors.New("createCommitFromDiff committer name and email are required when provided")
		}
		options.Committer.Name = strings.TrimSpace(options.Committer.Name)
		options.Committer.Email = strings.TrimSpace(options.Committer.Email)
	}

	return options, nil
}

func (d *diffCommitExecutor) send(ctx context.Context, repoID string) (CommitResult, error) {
	options, err := d.normalize()
	if err != nil {
		return CommitResult{}, err
	}

	ttl := resolveCommitTTL(options.InvocationOptions, defaultTokenTTL)
	jwtToken, err := d.client.generateJWT(repoID, RemoteURLOptions{Permissions: []Permission{PermissionGitWrite}, TTL: ttl, RefPolicies: options.RefPolicies})
	if err != nil {
		return CommitResult{}, err
	}

	diffReader := options.Diff
	if diffReader == nil {
		return CommitResult{}, errors.New("createCommitFromDiff diff is required")
	}

	metadata := buildDiffCommitMetadata(options)

	pipeReader, pipeWriter := io.Pipe()
	encoder := json.NewEncoder(pipeWriter)
	encoder.SetEscapeHTML(false)

	go func() {
		defer pipeWriter.Close()
		if err := encoder.Encode(metadataEnvelope{Metadata: metadata}); err != nil {
			_ = pipeWriter.CloseWithError(err)
			return
		}
		if err := writeDiffChunks(encoder, diffReader); err != nil {
			_ = pipeWriter.CloseWithError(err)
			return
		}
	}()

	requestURL := d.client.api.basePath() + "/repos/" + url.PathEscape(repoID) + "/diff-commit"
	resp, err := doStreamingRequest(ctx, d.client.api.httpClient, http.MethodPost, requestURL, jwtToken, pipeReader)
	if err != nil {
		return CommitResult{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		fallback := "createCommitFromDiff request failed (" + itoa(resp.StatusCode) + " " + resp.Status + ")"
		statusMessage, statusLabel, refUpdate, err := parseCommitPackError(resp, fallback)
		if err != nil {
			return CommitResult{}, err
		}
		return CommitResult{}, newRefUpdateError(statusMessage, statusLabel, refUpdate)
	}

	var ack commitPackAck
	if err := decodeJSON(resp, &ack); err != nil {
		return CommitResult{}, err
	}

	return buildCommitResult(ack)
}

func buildDiffCommitMetadata(options CommitFromDiffOptions) *commitMetadataPayload {
	metadata := &commitMetadataPayload{
		TargetBranch:  options.TargetBranch,
		CommitMessage: options.CommitMessage,
		Author: authorInfo{
			Name:  options.Author.Name,
			Email: options.Author.Email,
		},
	}

	if options.ExpectedTargetSHA != "" {
		metadata.ExpectedTargetSHA = options.ExpectedTargetSHA
	}
	if options.BaseBranch != "" {
		metadata.BaseBranch = options.BaseBranch
	}
	metadata.TargetIsEphemeral = options.TargetIsEphemeral
	metadata.BaseIsEphemeral = options.BaseIsEphemeral
	if options.Committer != nil {
		metadata.Committer = &authorInfo{
			Name:  options.Committer.Name,
			Email: options.Committer.Email,
		}
	}

	return metadata
}

func writeDiffChunks(encoder *json.Encoder, reader io.Reader) error {
	buf := make([]byte, maxChunkBytes)
	var pending []byte
	for {
		n, err := reader.Read(buf)
		if n > 0 {
			if pending != nil {
				payload := diffChunkEnvelope{
					DiffChunk: diffChunkPayload{
						Data: base64.StdEncoding.EncodeToString(pending),
						EOF:  false,
					},
				}
				if err := encoder.Encode(payload); err != nil {
					return err
				}
			}
			pending = append(pending[:0], buf[:n]...)
		}
		if err == io.EOF {
			if pending == nil {
				payload := diffChunkEnvelope{
					DiffChunk: diffChunkPayload{
						Data: "",
						EOF:  true,
					},
				}
				return encoder.Encode(payload)
			}
			payload := diffChunkEnvelope{
				DiffChunk: diffChunkPayload{
					Data: base64.StdEncoding.EncodeToString(pending),
					EOF:  true,
				},
			}
			return encoder.Encode(payload)
		}
		if err != nil {
			return err
		}
	}
}

func normalizeDiffBranchName(value string) (string, error) {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return "", errors.New("createCommitFromDiff targetBranch is required")
	}
	if strings.HasPrefix(trimmed, "refs/heads/") {
		branch := strings.TrimSpace(strings.TrimPrefix(trimmed, "refs/heads/"))
		if branch == "" {
			return "", errors.New("createCommitFromDiff targetBranch must include a branch name")
		}
		return branch, nil
	}
	if strings.HasPrefix(trimmed, "refs/") {
		return "", errors.New("createCommitFromDiff targetBranch must not include refs/ prefix")
	}
	return trimmed, nil
}
